### ISSUE TITLE
XML catalog support apparently not working

### ISSUE THREAD

**Issue by ralfhandl** [redhat-developer/vscode-xml#78]
I need to validate XML files against their XML schemas, and injecting xsi:schemaLocations into those files is not possible/feasible.

XML catalogs should solve that, yet I can't get that to work.

What I tried in USER SETTINGS:
```json
    "xml.catalogs": [
        "O:\\git\\odata-csdl-schemas\\catalog.xml"
    ]
```

The file `O:\git\odata-csdl-schemas\catalog.xml` contains
```xml
<?xml version="1.0"?>
<catalog xmlns="urn:oasis:names:tc:entity:xmlns:xml:catalog">
  <uri name="urn:oasis:names:tc:entity:xmlns:xml:catalog" uri="jar:file:/C:/eclipse-Photon/plugins/org.eclipse.wst.standard.schemas_1.0.800.v201711201732.jar!/xsd/xmlcatalog11.xsd" />
  <uri name="http://docs.oasis-open.org/odata/ns/edmx" uri="http://localhost/odata-schemas/edmx.xsd" />
  <uri name="http://docs.oasis-open.org/odata/ns/edm" uri="http://localhost/odata-schemas/edm.xsd" />
</catalog>
```

The first entry is copied from Eclipse's system_catalog.xml and should help validate the catalog.xml - it doesn't.

The second and third entries map the XML namespaces I'm actually interested in to local copies of the corresponding XML Schema files. The files are accessible via my local web server at the URIs listed in the file. Neither validation nor code completion are working as expected.

Removing the first entry with the `jar:` URL doesn't make the other two work.

What am I doing wrong?

Thanks in advance
Ralf

**Comment by fbricon** [comment]
@angelozerr seems like catalog support is broken in master. Testing the good old IBM examples https://raw.githubusercontent.com/microclimate-devops/xml-language-server/master/client-vscode/xml-client/readme-resources/examples.zip, I get no errors, but that diagnostic that won't go away:
<img width="679" alt="screen shot 2018-11-05 at 11 55 37 am" src="https://user-images.githubusercontent.com/148698/48013215-d0d28d80-e0f1-11e8-99ff-70172c9da825.png">

I set 
```json
    "xml.catalogs": [
        "catalog/catalog.xml"
    ],
```

**Comment by angelozerr** [comment]
> - however that does not fix the problem.

Ok thanks for your feedback, I must find time to study your usecase.I will ping you when if I find and fixes problem.

> Not sure if that is what you mean by "upstream"

I meant using last build of vscode-xml, sorry with my bad explanation.

**Comment by ralfhandl** [comment]
Thanks!

Do you have a working example using a catalog?

**Comment by angelozerr** [comment]
For our lsp4xml tests, we use https://github.com/angelozerr/lsp4xml/blob/master/org.eclipse.lsp4xml/src/test/resources/catalogs/catalog.xml

**Comment by ralfhandl** [comment]
Thanks. Can you point me to an XML file that is successfully validated using this catalog.xml?

**Comment by angelozerr** [comment]
We have not XML file in our test, just string, see for instance https://github.com/angelozerr/lsp4xml/blob/master/org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaDiagnosticsTest.java#L31

**Comment by fbricon** [comment]
@ralfhandl you can also use file association instead, to bind some xml file patterns to schemas, so you don't rely on xsi:schemaLocation or xsi:noNamespaceSchemaLocation. 

It'd be easier if you provided a set of sample documents you want to associate (through catalog or other mechanism).

**Comment by fbricon** [comment]
> so I followed the instructions in https://github.com/redhat-developer/vscode-xml/blob/master/CONTRIBUTING.md but npm run build-server fails both on plain Windows 10 and on "Linux on Windows".

@ralfhandl can you please open an issue for that? If you could provide more details about that failure, that'd help.

**Comment by ralfhandl** [comment]
Here's what I want to do:
* my XML documents declare two XML namespaces, one with prefix `xmlns:edmx=...`, one as default namespace without prefix `xmlns=...`
* `catalog.xml` maps the XML namespace URLs to XSD files in folder `schemas` of this workspace
* the workspace settings reference `catalog.xml` in the workspace root
* folder `examples` contains XML documents, here only `mini-catalog.xml`. This file contains two XML comments preceding empty lines and telling what suggestions I expect in those empty lines

Here's a ZIP with an example workspace: [vscode-xml-catalog-test.zip](https://github.com/redhat-developer/vscode-xml/files/2558034/vscode-xml-catalog-test.zip)

Note: the expected suggestions show fine in Eclipse Photon with XML Catalog entries mapping the two XML namespaces to files on my machine, so `mini-catalog.xml` is correct.

Things that I might have gotten wrong:
* relative path to `catalog.xml` in workspace settings
* relative paths to XSD files in `catalog.xml`
* syntax of `catalog.xml` (seems ok according to the XML Catalog specification and examples I found)
* ...

Thanks in advance
Ralf

**Comment by angelozerr** [comment]
@ralfhandl thank a lot for your great sample and explanation. XML catalog configuration is a nightmare and without debugging the lsp4xml, I was not able to know your problems.

To be honnest with you, I'm not a big expert with XML catalog, but here steps to follow to using yoru catalog:

 * the xml.catalog settings must use the full path. For the moment you must use full path. I'm working in this issue at https://github.com/angelozerr/lsp4xml/issues/206
 * in your catalog.xml, I don't know why, you must write:

`<public publicId="http://docs.oasis-open.org/odata/ns/edmx" uri="./schemas/edmx.xsd" />`

and not 

`<uri name="http://docs.oasis-open.org/odata/ns/edmx" uri="./schemas/edmx.xsd" />`

I must study this problem.

As XML catalog write and configuration is an hell, we should provide some support like:

 * https://github.com/angelozerr/lsp4xml/issues/205
 * https://github.com/angelozerr/lsp4xml/issues/204

Don't hesitate to create other issues to improve catalog support in lsp4xml.

**Comment by ralfhandl** [comment]
@angelozerr thanks for the investigation!

A few things I noted:

1. Full path to catalog has to use syntax `"c:/path/to/catalog.xml"`. Neither `"file:/c:/...`" nor `"file:///c:/...`" nor `"c:\\...`" worked
2. `catalog.xml` has to include `<public publicId=...` for the root XSD _plus_ one `<uri name=...` for the root XSD and each imported XSD, in my case:
```xml
<public publicId="http://docs.oasis-open.org/odata/ns/edmx" uri="./schemas/edmx.xsd" />
<uri name="http://docs.oasis-open.org/odata/ns/edmx" uri="./schemas/edmx.xsd" />
<uri name="http://docs.oasis-open.org/odata/ns/edm" uri="./schemas/edm.xsd" />
```
3. Having no `<uri name=...` entry for the root XSD leads to the validation message `[xml] cvc-elt.1.a: Cannot find the declaration of element 'edmx:Edmx'.`
4. This seems to be tied to whether the XML document declares the XML namespace with prefix or not: namespaces declared with `xmlns:abc=...` need two catalog entries: a public and a uri entry. The namespace declared as default with just `xmlns=...` only needs the uri entry in the catalog. Attached new zip file [vscode-xml-catalog-test.zip](https://github.com/redhat-developer/vscode-xml/files/2562459/vscode-xml-catalog-test.zip) with `examples/mini-catalog-prefixed.xml` and a `catalog.xml` without public entry for the second namespace: code completion shows nothing for the namespace prefix `edm:` without public entry
5. Code completion sometimes adds the wrong namespace prefix to an element from an included XSD, and sometimes some elements appear twice. Here `Annotation` should not have a prefix as it is defined in the default namespace, and `edmx:Include` should appear only once:
![image](https://user-images.githubusercontent.com/951576/48209470-8eb97e00-e375-11e8-94db-d4d2f4a84141.png)

**Comment by angelozerr** [comment]
@ralfhandl thank again for your feedback.

If you reinstall the last build of vscode-xml, it should supports:

 * "./catalog.xml" syntax
 * "file:///C:/catalog.xml"

for that see https://github.com/angelozerr/lsp4xml/issues/206

If you enable log, you could see if your catalog path is correct or not https://github.com/angelozerr/lsp4xml/issues/205#issuecomment-437250126

I continue to investigate your other problems.

**Comment by angelozerr** [comment]
@ralfhandl 2), 3), 4) are linked to https://github.com/angelozerr/lsp4xml/issues/217 that I have tried to fix.

If https://github.com/angelozerr/lsp4xml/issues/217 is working I think we can close this issue, no?

**Comment by ralfhandl** [comment]
@angelozerr The main issue is fixed, and the sub-topics are addressed by other issues, so feel free to close this issue.

Thanks again for the quick solution!

**Comment by angelozerr** [comment]
You have some troubles with XML catalog now? If yes, which troubles?

**Comment by ralfhandl** [comment]
@angelozerr XML catalog support works fine now with build #506, and this issue can be closed.

I have run into a minor new issue with completion when using namespace prefixes for both namespaces. I've opened https://github.com/redhat-developer/vscode-xml/issues/87 for this separate issue.
